pandas/core/indexes/interval.py

```diff
@@ -1038,7 +1038,16 @@ def overlaps(self, other):
 
     def _setop(op_name):
         def func(self, other):
-            other = self._as_like_interval_index(other)
+            try:
+                other = self._as_like_interval_index(other)
+            # allow ValueError from this method to raise to catch mixed closed
+            # except only Non-Interval index mismatches.
+            except TypeError:
+                # Currently this will cause difference operations to return
+                # object dtype as opposed to IntervalIndex, unlike other Index
+                # objects that return the same type when using `difference` on
+                # mismatched types
+                return getattr(self.astype('O'), op_name)(other)
```

Comment:
To resolve your comment, I think you could do something like:
```python
result = getattr(self.astype('O'), op_name)(other)
if op_name in ('intersection', 'difference'):
    result = result.astype(self.dtype)
return result
```
Note that I added `'intersection'` to the `if` statement, as it makes sense to me that `intersection` should return the same type as the calling index.  This does not appear to be handled consistently though, so wouldn't be opposed to going in the other direction there:
```python
In [2]: dti = pd.date_range('2018', periods=3); dti
Out[2]: DatetimeIndex(['2018-01-01', '2018-01-02', '2018-01-03'], dtype='datetime64[ns]', freq='D')

In [3]: fidx = pd.Float64Index([1.1, 2.2, 3.3]); fidx
Out[3]: Float64Index([1.1, 2.2, 3.3], dtype='float64')

In [4]: dti.intersection(fidx)
Out[4]: DatetimeIndex([], dtype='datetime64[ns]', freq=None)

In [5]: fidx.intersection(dti)
Out[5]: Index([], dtype='object')
```

I wanted to say that you could just special case things based on the indexes being disjoint at this point in time, but I guess there could be a weird corner case like `pd.Index(['foo', pd.Interval(0, 1)])` that should technically be handled, so the current approach seems like it is the most complete.  I think the mixed closed interval case could be handled in the same manner?

Could probably do some clever things to make this more performant, but that seems out of scope for this PR, and in general these methods could use a rewrite for perf given that the valid interval-interval case does a roundtrip through `MultiIndex`.